internal/graph/checkingresourcestream.go

```diff
@@ -0,0 +1,537 @@
+	defer rq.lock.Unlock()
+
+	toProcess := maps.Values(rq.toProcess)
+	if !alwaysReturn && len(toProcess) < int(datastore.FilterMaximumIDCount) {
+		return nil
+	}
+
+	for _, pr := range toProcess {
+		rq.beingProcessed[pr.orderingIndex] = pr
+	}
+
+	rq.toProcess = map[uint64]possibleResource{}
+	return toProcess
+}
+
+// resourcesToPossiblyPublish returns all resources in the toPublish state. This does *not* mark the resources
+// as published.
+func (rq *resourceQueue) resourcesToPossiblyPublish() []possibleResource {
+	rq.lock.Lock()
+	defer rq.lock.Unlock()
+
+	return maps.Values(rq.toPublish)
+}
+
+// checkingResourceStream is a Stream[*v1.DispatchLookupResourcesResponse] that consumes reachable resource
+// responses which are published to it, checks the resource (if necessary), and then publishes the resource
+// if reachable. This stream performs Checks for the possibly-inaccessible resources in a parallel fashion
+// but maintains the proper publishing order to the parent stream.
+//
+// Resources in the stream are processed as follows:
+//  1. A reachable resource is published to the stream via a call to the Publish method
+//  2. The resource is placed into the resourceQueue with an index indicating its publishing order
+//  3. A processing worker (up to concurrency limits) grabs the resources to be processed in the queue, checks
+//     those resources that need to have CheckPermission invoked, and places all resources processed in the queue
+//     into the "ready for publishing" state.
+//  4. The *single* publishing worker grabs resources to be published and publishes thme in the correct ordering,
+//     skipping any resources whose CheckPermission calls showed them as being inaccessible.
+//  5. The waitForPublishing call waits for the stream to have fully processed and published all queued resources
+//     before returning.
+type checkingResourceStream struct {
+	// ctx is the parent context for the LookupResources.
+	// NOTE: This will be disconnected from the reachableContext below.
+	ctx    context.Context
+	cancel func()
+
+	// reachableContext is the context to be returned by this stream for the ReachableResources call, and is
+	// disconnected from the overall context.
+	reachableContext context.Context
+
+	// concurrencyLimit is the limit on the number on concurrency processing workers.
+	concurrencyLimit uint16
+
+	req          ValidatedLookupResourcesRequest
+	checker      dispatch.Check
+	parentStream dispatch.Stream[*v1.DispatchLookupResourcesResponse]
+
+	// sem is a chan of length `concurrencyLimit` used to ensure the task runner does
+	// not exceed the concurrencyLimit with spawned goroutines.
+	sem chan token
+
+	// rq is the resourceQueue for managing the state of all resources returned by the reachable resources call.
+	rq *resourceQueue
+
+	// availableForProcessing is a channel which indicates to the processing worker(s) that work is available
+	// for processing.
+	availableForProcessing chan struct{}
```

Comment:
maybe `reachableResourceAvailable`? Or maybe turn the one below 👇🏻 into `processingCompleted`. Suggesting because by reading the code I was confused by which phase was completed through the `select` loops